Posts/2021/09/tfl-profile.md

```diff
@@ -0,0 +1,126 @@
+
+**A:** Profile files are a powerful part of PowerShell and allow you to customize PowerShell for your environment.
+They are easy to create and support a range of deployment scenarios.
+
+## What is a Profile File?
+
+Before explaining the profile file, let's first examine the PowerShell host.
+A PowerShell host is a program that hosts PowerShell to allow you to use it.
+Common PowerShell hosts include the Windows PowerShell console, the Windows PowerShell ISE, the PowerShell 7 console, and VS Code.
+Each host supports the use of profile files.
+
+A profile file is a PowerShell script that a PowerShell host loads and executes automatically every time you start that PowerShell host.
+The script is, in effect, dot-sourced, so any variables, functions, and the like that you define in a profile script remain available in the PowerShell session, which is incredibly handy.
+I use profiles to create PowerShell drives, various useful variables, and a few useful (for me!) functions.
+
+Each PowerShell host has 4 separate profile files as follows:
+
+* This host, this user
+* This host, all users
+* All hosts, this user
+* All hosts, all users
+
+Why so many, you might ask.
+Because having these four profile files allows you numerous deployment opportunities.
+You could, for example, have one profile file for every PowerShell host on a machine for all users that defines corporate aliases or standard PS drives.
+You could have 'this host' profiles that define host-specific customizations which could differ depending on the PowerShell host.
+For example, in my profile file for VS code, I use ``Set-PSReadLineOption`` to set token colours depending on which color theme I am using.
+Like so many things in PowerShell, the PowerShell team engineered profiles for every scenario you might come across in deploying PowerShell and PowerShell hosts.
+
+In practice, the "this host, this user" profile is the one you most commonly use, but having all four allows considerable deployment flexibility.
+You have options!
+
+## Where do I find them?
+
+Another frequently asked questions are: where are these files and how are they named?
+It turns out, like many things PowerShell, you can find the answer to the question inside PowerShell itself.
+In this case, inside a PowerShell automatic variable, **$PROFILE**.
+
+Automatic variables in PowerShell, are variables created by PowerShell itself and are available for use.
+These variables are created by the host when you start up the host.
+For more details on automatic variables see the [automatic variable help text](https://docs.microsoft.com/powershell/module/microsoft.powershell.core/about/about_automatic_variables).
+
+## The **$PROFILE** variable
+
+The **$PROFILE** variable is an automatic variable that the PowerShell host creates within each session during startup.
+This variable has both a **ToString()** method and four additional note properties that tell you where __this__ host finds its profile files.
```

Comment:
Did you mean to bold "this"? For bold use two asterisks. For italic using single underscores.

```suggestion
This variable has both a **ToString()** method and four additional note properties that tell you where _this_ host finds its profile files.
```